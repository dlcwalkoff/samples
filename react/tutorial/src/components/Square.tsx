import React = require("react");

export interface SquareBoard {
    value: string,
    onClick: () => void
}

export class Square extends React.Component {
    constructor(public props: SquareBoard) {
        super(props);
    }

    render() {
        return (
            <button className='square' onClick={() => this.props.onClick()}>
                {this.props.value}
            </button>
        );
    }
}