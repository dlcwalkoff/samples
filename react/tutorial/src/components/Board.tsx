import * as React from 'react';
import { Square } from './Square';
import { Winner, SquareValue } from '../classes/Winner';

export interface BoardState {
    squares: Array<SquareValue>,
    xIsNext: boolean,
}

export class Board extends React.Component {
    public state: BoardState = {
        squares: Array(9).fill(null),
        xIsNext: true,
    }

    constructor(props: any) {
        super(props);
    }

    handleClick = (i: number) => {
        const squares = [...this.state.squares];

        if (Winner.calculate(squares) || squares[i]) {
            return;
        }

        squares[i] = this.state.xIsNext ? 'X' : 'O';

        this.setState({ squares, xIsNext: !this.state.xIsNext, });
    };

    renderSquare(i: number) {
        return <Square value={this.state.squares[i]} onClick={() => this.handleClick(i)} />;
    }

    render() {
        const winner = Winner.calculate(this.state.squares);
        let status;

        if (winner) {
            status = 'Winner: ' + winner;
        } else {
            status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
        }
    
        return (
          <div>
            <div className="status">{status}</div>
            <div className="board-row">
              {this.renderSquare(0)}
              {this.renderSquare(1)}
              {this.renderSquare(2)}
            </div>
            <div className="board-row">
              {this.renderSquare(3)}
              {this.renderSquare(4)}
              {this.renderSquare(5)}
            </div>
            <div className="board-row">
              {this.renderSquare(6)}
              {this.renderSquare(7)}
              {this.renderSquare(8)}
            </div>
          </div>
        );
    }
}