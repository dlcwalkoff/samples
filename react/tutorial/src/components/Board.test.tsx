import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React = require('react');
import { Board } from './Board';
import { Square } from './Square';


describe('Board', () => {
  it('should render correctly', () => {
    const output = shallow(
      <Board />
    );

    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should render 9 Squares', () => {
    const output = shallow(
      <Board />
    );

    expect(output.find(Square)).toHaveLength(9);
  });

  it('should right change state after Square click', () => {
    const output = shallow(
      <Board />
    );

    expect(output.state('squares')).toEqual([null, null, null, null, null, null, null, null, null]);
    expect(output.state('xIsNext')).toBe(true);

    output.find(Square).first().simulate('click');

    expect(output.state('squares')).toEqual(['X', null, null, null, null, null, null, null, null]);
    expect(output.state('xIsNext')).toBe(false);

    output.find(Square).at(1).simulate('click');

    expect(output.state('squares')).toEqual(['X', 'O', null, null, null, null, null, null, null]);
    expect(output.state('xIsNext')).toBe(true);

  });
});