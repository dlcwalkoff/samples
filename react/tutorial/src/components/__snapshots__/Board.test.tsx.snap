// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should render correctly 1`] = `
<div>
  <div
    className="status"
  >
    Next player: X
  </div>
  <div
    className="board-row"
  >
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
  </div>
  <div
    className="board-row"
  >
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
  </div>
  <div
    className="board-row"
  >
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
  </div>
</div>
`;
